crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3729,13 +3723,19 @@ impl<'db> TupleType<'db> {
     ) -> Type<'db> {
         let mut elements = vec![];
 
+        let mut is_never = false;
         for ty in types {
             let ty = ty.into();
             if ty.is_never() {
-                return Type::Never;
+                // We can not return early here because we need to consume the whole iterator,
+                // to ensure that all element types are inferred.
+                is_never = true;
```

Comment:
I'm not a 100% if here is the right place to fix this because `from_elements` could also be used from non-inference code. Maybe it's better to keep the collect in `infer.rs`? 

If you decide to keep it here. I then suggest that we avoid copy over the `elements` if we've seen any `Never` type
```suggestion
                // We can not return early here because we need to consume the whole iterator,
                // to ensure that all element types are inferred.
                let _ = types.last();
                return Type::Never
```

This probably requires using a `while let`

Reply:
> I'm not a 100% if here is the right place to fix this because `from_elements` could also be used from non-inference code.

To be clear: I'm assuming you are only referring to a potential performance implication — not some correctness argument?

My thinking was that I'd rather remove this footgun alltogether as opposed to squeezing out a tiny bit of performance: when do we really gain something — performance-wise — by returning early here? Only if we have a very long tuple type where one of the first elements is `Never`. This is an extremely unlikely case that I don't think is worth optimizing for?

Reply:
> I'm not a 100% if here is the right place to fix this because `from_elements` could also be used from non-inference code. Maybe it's better to keep the collect in `infer.rs`?

I agree, I think it's best to keep the collect in `infer.rs` and keep the early return here. It seems better for a clean separation of concerns.

Could we also add an mdtest that reproduces the regression that was nearly introduced here? That way we won't accidentally lose test coverage if the fixture file in `crates/ruff_linter` is changed for some unrelated reason

Reply:
> My thinking was that I'd rather remove this footgun alltogether

That's a reasonable point... it just still feels to me like it's the wrong place to be fixing the footgun. Conceptually, the code in `types.rs` should be "pure code" that operates on types and produces other types, rather than code that is concerned with inferring types from ASTs

Reply:
That's a fair point. An alternative would be to change the signature to accept a `&[Type<'db>]` which, would remove the need to collecting the types and remove `from_elements` to a simple check if the slices contains `Never`.

```
		if types.iter().any(Type::is_never) {
            return Type::Never;
        }

        Type::Tuple(Self::new(db, elements))
```

Reply:
I saw this too late and changed it back to using `.collect` in `infer.rs`. I added a comment and a regression test. I also manually reviewed all other occurrences of `TupleType::from_elements` and I *think* they are not problematic. But of course, there is no guarantee that we wont run into this again.

I do fully agree with your point that this has actually nothing to do with `TupleType::from_elements` but is more of a general problem in `infer.rs`. We need to be careful everywhere, and I was not, when removing the `.collect` here.